README
Full online documentation at:
https://wiki-bsse.ethz.ch/display/ShoRAH/Documentation

ShoRAH consists of several programs:

dec.py       - local error correction based on diri_sampler
diri_sampler - Gibbs sampling for Dirichlet process mixture
contain.cc   - eliminate redundant reads
mm.py        - maximum matching haplotype construction
freqEst.cc   - EM algorithm for haplotype frequency
fas2read.pl  - translates between two formats for read data files
s2f.py       - multiple sequence alignment of NGS reads
bam2msa.py   - extracts a multiple sequence alignment from a BAM file

Copyright 2007, 2008, 2009, 2010
Niko Beerenwinkel,
Arnab Bhattacharya,
Nicholas Eriksson,
Moritz Gerstung,
Lukas Geyrhofer,
Osvaldo Zagordi,
ETH Zurich
under GPL. See file LICENSE for details.
The program also includes other software written by third parties.
This has been distributed according to the licenses provided by the
respective authors.


GENERAL USAGE:
--------------------------------------------------

Install:
	type 'make' to build the C++ programs, then run. See INSTALL
	for additional information.

Run:
	The whole process can be run one step after the other, or one can
	invoke shorah.py, that runs the whole process from read	file to
	frequency estimation. The local analysis alone can be run invoking
	dec.py or directly diri_sampler. The input must be a fasta file 	with reads or a multiple alignment of the reads in fasta format
	(.far extension). Shorah uses the program s2f to produce the
	alignment, or the user copy its alignment to a file with the .far
	extension and run shorah on it. We also provide a program to
	extract a multiple sequence alignment from a BAM file.



S2F.PY
--------------------------------------------------
Options:
  -h, --help            show this help message and exit
  -f INPUT, --readfile=INPUT
  -r REF, --ref=REF     
  -o O, --output=O      output suffix must be '.far' or none for stdout
  -t THRESHOLD, --threshold=THRESHOLD
                        if similarity is less, throw reads away...
                        <default=0.7>
  -d, --delete_files    delete temporary files <default=False>
  -n, --no_pad_insert   do not insert padding gaps <default=insert>

Output:
	The program s2f.py takes as input the file of reads (in fasta
	format) and the	reference genome. It produces a set of pairwise
	alignments, then merge them to obtain a multiple one in a file
	with far extension (file_name.far). This is the input to the
	subsequent analysis performed with dec.py.

Examples:
	A sample file has been included for testing purposes,
	sample_454.fasta and the reference ref_genome.fasta.
	sample_454.fasta consists of 1000 reads of mean length equal
	to 310 bp, derived from two haplotypes mixed in 80:20
	proportion.
	We recommend moving both files to a different working directory
	and follow the rule of thumb: "one experiment in one directory".
	
	We assume shorah was downloaded in the directory
	path_to_sho_bin/. The far file can be created  from the fasta file
	and the reference by invoking
	[user@host]$ path_to_sho_bin/shorah-0.4/s2f.py -f sample_454.fasta -r ref_genome.fasta -o sample_454.far

	





DEC.PY:
--------------------------------------------------
Options:
  -h, --help            show this help message and exit
  -f F, --readsfile=F   file with reads <.far format>
  -j J, --iterations=J  iterations in dpm sampling <2000>
  -a A, --alpha=A       alpha in dpm sampling <0.01>
  -w W, --windowsize=W  window size <201>
  -s S, --winshifts=S   number of window shifts <3>
  -k, --keep_files      keep all intermediate files of diri_sampler
                        <default=False>
  -r REF, --ref=REF     reference genome
  -t THRESHOLD, --threshold=THRESHOLD
                        if similarity is less, throw reads away...
                        <default=0.7>
  -d, --delete_s2f_files
                        delete temporary align files of s2f.py <default=False>
  -n, --no_pad_insert   do not insert padding gaps in .far
                        file<default=insert>


Other options
-------------
These parameters are controlled by the global variables declared at the beginning of file dec.py. See the relative comments for an explanation.
IMPORTANT: Since version 0.3 dec.py runs in parallel on multi-processor
computers. By default it uses all available CPUs, if you want to limit the
number of parallel processes, edit the variable max_proc in dec.py

Tips:
	Rerunning on selected windows:
	When running with option -k, dec.py produces and retains several
	files in the subdirectory "corrected/" of the form
	w1-99.reads-cor.fas.gz, where 1 and 99 are, respectively, the
	beginning and the end of the window under consideration. If the
	user wishes to rerun dec.py on selected windows only, one should
	simply delete the corresponding file in the directory "corrected/"
	NOTE: The error correction implemented in diri_sampler will not run 	for any window whose corresponding file is present in the directory
	"corrected/".
	
	Changing parameters on the fly:
	If the user wishes to change the number of iterations of
	diri_sampler (because the burn-in phase is shorter or longer
	than expected) one can write a file w1-99.iter (again 1 and 99
	are the boundaries of the window) with a single line that
	contains the desired number of iterations. Similarly, one can
	change alpha by writing into a file w1-99.alpha a float with the
	new value of alpha. While alpha can be increased and decreased by
	editing the .alpha file without adversely affecting the
	functioning of the program (but obviously its result), we
	recommend changing the number of iterations at most once.

	Maximum weight of read objects:
	The new implementation of diri_sampler clubs identical reads into
	objects defined by the read sequence and the number of reads in
	the object. This allows for faster computation, but might
	introduce a bias into the sampling procedure. One can limit the
	maximum number of identical reads in a single object by	editing
	the variable LIMIT in dpm_src/dpm_sampler.h (default is 10000)
	
Output:
	The program dec.py outputs several files. The file sample_cor.fas,	which contains the corrected reads, is passed by shorah (or by the	user) to the following step of the global analysis, fas2read.pl.	dec.py also outputs a file called proposed.dat. It contains the
	number of proposed new clusters in each window. One should always	make sure that a sufficient number of new cluster has been proposed 	(rule of thumb, at least 0.1 per step or more). When the program is 	run with the -k option, several directories are created. They
	contain zipped files that report information on the error
	correction procedure. For a full explanation, see
	http://wiki-bsse.ethz.ch/display/ShoRAH/Local+analysis



SHORAH.PY:
--------------------------------------------------


Options:
  -h, --help            show this help message and exit
  -f F, --readsfile=F   file with reads <.fas or .far format>
  -j J, --iterations=J  iterations in dpm sampling <1000>
  -a A, --alpha=A       alpha in dpm sampling <0.01>
  -w W, --windowsize=W  window size in <201>
  -t THRESHOLD, --threshold=THRESHOLD
                        if similarity is less, throw reads away...
                        <default=0.7>
  -n, --no_pad_insert   do not insert padding gaps <default=insert>
  -r REF, --ref=REF     
  -s S, --winshifts=S   number of window shifts <3>
  -k, --keep_files      keep intermediate files (Gibbs sampling)
                        <default=False>


Examples:
	How to run shorah on the example file described above.
	[user@host]$ path_to_sho_bin/shorah-0.4/shorah.py -f sample_454.fasta -r ref_genome.fasta -j 1000 -w 90 -a 0.1 -k &> global.log
	

	
Output:
	The final result of the analysis is in the file
	sample_454_cor.popl, the first lines of which are reported.
	>HAP0_0.708656
	TC-AA--A--TCACTCTTTGGCAACGACCCCTTGTC-AC.........[line truncated]
	>HAP1_0.194
	TC-TG--A--TCACTCTTTGGCAGCGACCCCTCGTC-AC.........[line truncated]

	This file contains the sequences of the reconstructed haplotypes
	(reconstructed by the program mm.py) and their frequencies
	(estimated by the program freqEst. The name of each sentence is in
	the format HAPn_freq where n is an ordinal number and freq is the
	frequency. Haplotypes are sorted in descending order according to
	their frequencies.


CHANGELOG
New in version 0.5.1
---------------------

+ bug fixes in bams2msa.py

+ default behaviour of s2f.py is now not to propagate gaps

New in version 0.5
---------------------

+ SAM/BAM format added! Users can now produce MSA files
  from their alignments in SAM format

+ README and INSTALL have been updated, and more
  information is available on the web, at the ShoRAH
  documentation page
  https://wiki-bsse.ethz.ch/display/ShoRAH/Documentation

New in version 0.4
---------------------

+ dpm_sampler now computes Hamming distance in
  time proportional to the distance rather than
  to the length of the sequences.

+ dpm_sampler now clubs identical reads into
  objects. This method achieves a significant speed-up
  with Illumina datasets. Up to 65000 Illumina reads were
  analysed in a single window with good results.

+ freqEst sorts the output according to the frequency


New in version 0.3
---------------------

+ dpm_sampler is now C++! Thanks to the structures
  defined in C++ libraries (map and multimap), we
  are able to run requiring much less memory

+ Parallelization! dec.py now calls diri_sampler using
  a pool of independent workers, exploiting all the
  available compuational power, as well as s2f.py does

+ The sampling has been improved, now we can have a more
  reliable estimate of the quality of our local
  haplotype reconstruction

+ The output of diri_sampler is now better organized
  (if -k is given, all intermediate files are saved in
  subdirectories of the current)

+ The alignment program runs now in linear time (with
  respect to the number of reads), and deals with indels
  in a more clever way

+ All python programs now use the logging module to
  write logs of their operations

+ plot_sampling.py and plot_stat.py can be used to produce
  graph showing the behaviour of the Gibbs sampling


New in version 0.2
---------------------

+ New method to assign the reads after the sampling

+ The alignment is now provided by a separate
  program (step2far.py), so that the user can input
  his/her own alignment and install EMBOSS only if
  really necessary

+ Fixed numerical bugs when dealing with very high or
  very low probabilities


shorah.py
#!/usr/bin/env python

# Copyright 2007, 2008, 2009
# Niko Beerenwinkel,
# Nicholas Eriksson,
# Moritz Gerstung,
# Lukas Geyrhofer,
# Osvaldo Zagordi,
# ETH Zurich

# This file is part of ShoRAH.
# ShoRAH is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# ShoRAH is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with ShoRAH.  If not, see <http://www.gnu.org/licenses/>.

import subprocess
import os
import os.path
import sys

import logging
import logging.handlers

# Make a global logging object.
sholog = logging.getLogger("ShorahLog")

# set logging level
sholog.setLevel(logging.DEBUG)
# This handler writes everything to a file.
LOG_FILENAME = './shorah.log'
h = logging.handlers.RotatingFileHandler(LOG_FILENAME, 'w', maxBytes=100000, backupCount=5)
f = logging.Formatter("%(levelname)s %(asctime)s %(funcName)s %(lineno)d %(message)s")
h.setFormatter(f)
sholog.addHandler(h)

dn = os.path.dirname(__file__)

def run_f2r(filein):
    """
    2. translate to read format
    output: filein.read
    """
    
    my_prog = 'perl -I %s %s' % (dn+'/perllib', os.path.join(dn, 'fas2read.pl'))
    my_arg  = " -fasta %s" % filein
    assert os.path.isfile(filein), 'File %s not found' % filein

    # runs the fasta to read conversion
    try:
        retcode = subprocess.call(my_prog + my_arg, shell=True)
        if retcode < 0:
            sholog.error(my_prog + my_arg)
            sholog.error("Child %s was terminated by signal" % my_prog, -retcode)
        else:
            sholog.debug("Child %s returned %i" % (my_prog, retcode))
    except OSError, ee:
        sholog.error("Execution of %s failed:" % my_prog, ee)
    
    return retcode


def run_contain(filein):
    """
    3. eliminate redundant reads
    output: filein.rest
    """
    
    my_prog = os.path.join(dn, 'contain')
    my_arg  = " -f %s" % filein
    assert os.path.isfile('%s.fas' % filein), 'File %s not found' % filein

    #eliminates redundant reads
    try:
        retcode = subprocess.call(my_prog + my_arg, shell=True)
        if retcode < 0:
            sholog.error(my_prog + my_arg)
            sholog.error("Child %s was terminated by signal" % my_prog, -retcode)
        else:
            sholog.debug("Child %s returned %i" % (my_prog, retcode))
    except OSError, ee:
        sholog.error("Execution of %s failed:" % my_prog, ee)
    
    return retcode


def run_mm(filein, max_hap=200):
    """
    4. run maximum matching, output up to 200 haplotypes
    output: sample.$nr.geno
    """
    my_prog = os.path.join(dn, 'mm.py')
    my_arg  = " %s %i" % ('%s.rest' % filein, max_hap)

    assert os.path.isfile('%s.rest' % filein), 'File %s not found' % filein
    
    #minimal coverage of the read graph
    try:
        retcode = subprocess.call(my_prog + my_arg, shell=True)
        if retcode < 0:
            sholog.error(my_prog + my_arg)
            sholog.error("Child %s was terminated by signal" % my_prog, -retcode)
        else:
            sholog.debug("Child %s returned %i" % (my_prog, retcode))
    except OSError, ee:
        sholog.error("Execution of %s failed:" % my_prog, ee)

    return retcode


def run_freqEst(filein):
    """
    5. run EM
    output: sample.$nr.popl
    """
    my_prog = os.path.join(dn, 'freqEst')
    my_arg  = " -f %s" % filein
    assert os.path.isfile('%s.rest' % filein), 'File %s not found' % filein

    #estimate frequencies
    try:
        retcode = subprocess.call(my_prog + my_arg, shell=True)
        if retcode < 0:
            sholog.error(my_prog + my_arg)
            sholog.error("Child %s was terminated by signal" % my_prog, -retcode)
        else:
            sholog.debug("Child %s returned %i" % (my_prog, retcode))
    except OSError, ee:
        sholog.error("Execution of %s failed:" % my_prog, ee)

    return retcode


def main():
    """ Only called if run not interactively
    """
    import optparse
    import shutil
    import dec
    import s2f
    
    # parse command line
    optparser = optparse.OptionParser()
    
    optparser.add_option("-f", "--readsfile", help="file with reads <.fas or .far format>",
                         default="", type="string", dest="f")
    optparser.add_option("-j", "--iterations", help="iterations in dpm sampling <1000>", default=1000,
                         type="int", dest="j")
    optparser.add_option("-a", "--alpha", help="alpha in dpm sampling <0.01>", default=0.01,
                         type="float", dest="a")
    optparser.add_option("-w", "--windowsize", help="window size in <201>", default=201,
                         type="int", dest="w")
    optparser.add_option("-t","--threshold", help="if similarity is less, throw reads away... <default=0.7>", type="float",
                         dest="threshold", default=0.7)
    optparser.add_option("-p","--pad_insert",help="insert padding gaps <default=not insert>", action="store_true", default=False,
                         dest="pad")
    optparser.add_option("-r","--ref", type="string", default="",
                         dest="ref")
    optparser.add_option("-s", "--winshifts", help="number of window shifts <3>", default=3,
                         type="int", dest="s") # window shiftings, such that each base is covered up to win_shifts times
    optparser.add_option("-k","--keep_files",help="keep intermediate files (Gibbs sampling) <default=False>", default=False,
                         action="store_true", dest="k")
    
    (options, args) = optparser.parse_args()
    
    sholog.info(' '.join(sys.argv))
    in_file = options.f
    keep_all_files = options.k
    step = options.w
    win_shifts = options.s
    iters = options.j
    alpha = options.a
    
    try:
        [in_stem, in_format] = [os.path.split(in_file)[1].split('.')[0], os.path.split(in_file)[1].split('.')[1]]
    except IndexError:
        print 'The input file must be filestem.format'
        sholog.error('The input file must be filestem.format')
        sys.exit()
        
    if in_format != 'far':
        ref_file = options.ref
        out_file = os.path.join(os.getcwd(), in_stem+'.far')
        thresh = options.threshold
        if thresh >= 1.0:
            sys.exit('invalid threshold')
        pad_insert = options.pad
        sholog.debug('running s2f.py')
        s2f.main(ref_file, in_file, out_file, thresh, pad_insert, keep_all_files)
    
        # run dec.py
        sholog.debug('running dec.py')
        dec.main(out_file, step, win_shifts, keep_all_files, iters, alpha)
    else:
        # run dec.py
        sholog.debug('running dec.py')
        dec.main(in_file, step, win_shifts, keep_all_files, iters, alpha)
        
    # copy file and run fas2reads
    shutil.move('%s.cor.fas' % in_stem, '%s_cor.fas' % in_stem)
    sholog.debug('running fas2reads')
    
    retcode = run_f2r('%s_cor.fas' % in_stem)
    if retcode is not 0:
        sholog.error('fas2reads did not return 0')
        sys.exit()
    
    # run contain
    retcode = run_contain('%s_cor' % in_stem)
    if retcode is not 0:
        sys.exit()
    
    # run mm.py (this might become a module)
    retcode = run_mm('%s_cor' % in_stem)
    if retcode is not 0:
        sys.exit()
    
    # run EM freqEst
    retcode = run_freqEst('%s_cor' % in_stem)
    if retcode is not 0:
        sys.exit()
    # goodbye

if __name__ == "__main__":
    main()
